### PR TITLE
Wip/hughsie/reverse split

### PR DESCRIPTION
**Pull request by hughsie**
Fair warning; this is already running on the LVFS due to a small amount of panic from a major vendor. I'm happy to fix things up before pushing to master obviously.

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -611,7 +653,7 @@ class Firmware(db.Model):
     __table_args__ = {'mysql_character_set': 'utf8mb4'}
 
     firmware_id = Column(Integer, primary_key=True, unique=True, nullable=False, index=True)
-    vendor_id = Column(Integer, ForeignKey('vendors.vendor_id'), nullable=False)
+    vendor_id = Column(Integer, ForeignKey('vendors.vendor_id'), nullable=False) # the OEM, not ODM
```
Comment (superm1) [inline app/models.py]: It would be better to rename this to `oem_id` and down below not part of this PR where it refers to vendor to be `oem`.

File: app/templates/firmware-affiliation.html [hunk 2]
```diff
@@ -0,0 +1,38 @@
+{% extends "default.html" %}
+{% block title %}Download Limits{% endblock %}
```
Comment (superm1) [inline app/templates/firmware-affiliation.html]: This title needs to be updated right?

File: app/templates/upload.html [hunk 3]
```diff
@@ -26,15 +26,30 @@ <h2>Upload Firmware</h2>
...
 <form action="/lvfs/upload" method="post" enctype="multipart/form-data" class="form">
+{% if affiliations|length %}
+  <div class="form-group row">
+    <label for="target" class="col-sm-2 col-form-label">Upload for vendor</label>
+    <div class="col-sm-10">
+      <select class="form-control" name="vendor_id" required>
+        <option value="{{g.user.vendor.vendor_id}}">{{g.user.vendor.display_name}}</option>
+{% for r in affiliations %}
+        <option value="{{r.vendor.vendor_id}}">{{r.vendor.display_name}}</option>
+{% endfor %}
```
Comment (superm1) [inline app/templates/upload.html]: So in my experience with ODMs:
If there are more than one option available here it's probably best to make this default to a blank option that needs to be set on every upload.  This will help to prevent FW ending up at wrong OEM account.

File: app/views_firmware.py [hunk 4]
```diff
@@ -309,6 +309,64 @@ def firmware_limit_add():
...
+        #if not vendor.get_affiliation_by_odm_id(v.vendor_id):
+        #    continue
+        vendors.append(v)
+
+    return render_template('firmware-affiliation.html', fw=fw, vendors=vendors)
+
+@app.route('/lvfs/firmware/<int:firmware_id>/affiliation/change', methods=['POST'])
+@login_required
+def firmware_affiliation_change(firmware_id):
+    """ Modifies the update urgency and release notes for the update """
```
Comment (superm1) [inline app/views_firmware.py]: Wrong comment here

File: app/views_vendor.py [hunk 5]
```diff
@@ -349,3 +349,82 @@ def vendor_user_add(vendor_id):
...
+        return redirect(url_for('.vendor_affiliations', vendor_id=vendor_id), 302)
+    if not 'vendor_id_odm' in request.form:
+        return _error_internal('No value')
+
+    # check if it already exists
+    vendor_id_odm = int(request.form['vendor_id_odm'])
+    for rel in vendor.affiliations:
+        if rel.vendor_id_odm == vendor_id_odm:
+            flash('Failed to add affiliate: Already a affiliation with that ODM', 'warning')
+            return redirect(url_for('.vendor_affiliations', vendor_id=vendor_id), 302)
```
Comment (superm1) [inline app/views_vendor.py]: I guess this is going to be a manual task for merging accidental instances of `Odm`, `Odm Ltd`, `Odm Inc` etc?

**Comment by superm1** [comment]
Something I think missing is updated documentation for "new" ODMs coming on board around how affiliates work.  Stuff that will land here: https://fwupd.org/vendors

File: app/templates/firmware-affiliation.html (hunk 2 above)
Comment (hughsie) [reply app/templates/firmware-affiliation.html]: Fixed, thanks.

File: app/templates/upload.html (hunk 3 above)
Comment (hughsie) [reply app/templates/upload.html]: So more like:?

screenshot from 2018-07-02 20-01-12

File: app/views_firmware.py (hunk 4 above)
Comment (hughsie) [reply app/views_firmware.py]: Fixed, thanks.

File: app/views_vendor.py (hunk 5 above)
Comment (hughsie) [reply app/views_vendor.py]: Yes, we can cross that bridge when we come to it :)

**Comment by hughsie** [comment]
> So something that stands out to me is that if an ODM can have an account that is QA

Does an ODM account need QA users? I guess it's not beyond imagination that a vendor can be an OEM and a ODM (e.g. I can think of a couple of examples already).

I don't know if the code matches my understanding, but surely it would just make sense that a ODM QA user should only be able to promote firmware owned by the ODM? That was my logical interpretation, although I'm not sure those protections exist in the code.

Just so I understand it correctly, this is the test I was thinking about:

* add oem1 vendor
* add oem2 vendor
* add odm1 vendor
* make odm1 affiliated with oem1
* make odm1 affiliated with oem2
* add alice@odm1.com to vendor odm1 as a QA user
* add bob@odm1.com to vendor odm1 as a normal user
* bob uploads foo.cab on behalf of vendor oem1 (vendor_id = oem1, user_id=bob)

Is alice is able to see the firmware that was uploaded by bob? I'd say no.
Is alice able to promote the firmware uploaded by bob? Again I think no.

**Comment by FireWolf74** [comment]
I have one question.
Can you do a GUI as well or can we get a list of the updates for each system?

File: app/templates/upload.html (hunk 3 above)
Comment (superm1) [reply app/templates/upload.html]: Yeah that looks good to me.

**Comment by superm1** [comment]
@FireWolf74 OT, please bring that discussion outside of this PR.

> Does an ODM account need QA users? I guess it's not beyond imagination that a vendor can be an OEM and a ODM (e.g. I can think of a couple of examples already).

Depends on how autonomously they operate from the OEM.  I'd say that is something that "should" be configurable because ODM can own the final QA but that's probably an OEM decision if they want to hand over that control.

>Is alice is able to see the firmware that was uploaded by bob? I'd say no.
> Is alice able to promote the firmware uploaded by bob? Again I think no.

Well I guess that's the confusing part.  Who is responsible to promote it to testing/stable then in this instance?  I guess only someone who is QA for `oem1`?  So that means that bob@odm1.com uploads the file and then someone at oem1 has to promote it (just like LVFS current workflow).

From the way you described this, it makes it sound like the only way an ODM QA user could promote files is if they are both an OEM and ODM.

**Comment by hughsie** [comment]
>From the way you described this, it makes it sound like the only way an ODM QA user could promote files is if they are both an OEM and ODM.

The LVFS doesn't actually distinguish between an ODM or an OEM, to it they are both just instances of Vendor. I think the only way an ODM QA user should be able to move to testing and stable is for the firmware to be "owned" by the ODM, so **never** when the firmware is uploaded "on behalf of" the OEM.

**Comment by hughsie** [comment]
> because ODM can own the final QA

In that case the ODM should upload firmware with the vendor_id owner of the ODM, and then when the firmware is QAd it can have the vendor re-assigned even in the stable remote. It might not even need to be "owned" by the OEM at all if the ODM is responsible for the QA and testing.

**Comment by hughsie** [comment]
> Something I think missing is updated documentation for "new" ODMs coming on board around how affiliates work

How about the last commit?

File: app/templates/vendors.html [hunk 6]
```diff
@@ -77,4 +77,32 @@ <h3 class="card-header list-group-item-info">Why does the LVFS project sign the
...
+  There are some OEMs where the ODM will be the entity responsible for
+  uploading the firmware to the LVFS.
+  The per-device QA can either be done by the OEM or the ODM as required.
+</p>
+<p>
+  The LVFS administrator can mark other vendors as <em>affiliates</em> of
+  other vendors.
+  This gives the ODM permission to upload firmware that is <em>owned</em> by the
+  OEM on the LVFS, and would appear in the OEM embargoed metadata.
+  The OEM QA team is also able to edit the update description, move the firmware
```
[the remark this answers is not in the document]
Comment (hughsie) [reply app/templates/vendors.html]: Works for me, thanks.

File: app/templates/user-admin.html [hunk 7]
```diff
@@ -5,6 +5,17 @@
...
+<div class="alert alert-warning mt-1" role="alert">
+  <strong>BE VERY CAREFUL EDITING THE ADMIN</strong>.
+</div>
+{% elif u.user_id == g.user.user_id %}
+<div class="alert alert-warning mt-1" role="alert">
+  <strong>Be careful editing your own profile:</strong>
+  If you remove attributes they may need to be re-added manually by the LVFS administrator.
+</div>
+{% endif %}
+
```
Comment (superm1) [inline app/templates/user-admin.html]: Rather than the heavily worded message, why not hide the modify button and mention it's read only?

File: app/models.py [hunk 8]
```diff
@@ -49,6 +49,7 @@ class User(db.Model):
     is_qa = Column(Boolean, default=False)
     is_analyst = Column(Boolean, default=False)
     is_vendor_manager = Column(Boolean, default=False)
+    is_approved_public = Column(Boolean, default=False)
```
Comment (superm1) [inline app/models.py]: I think i'd prefer to see `approved_promoter` to be clearer what this new permission level means.  I think it makes sense though.

Comment (hughsie) [reply app/models.py]: I went with public as we have a `Remote.is_public` to mean stable and testing. What I'm trying to say is "we know the user can legally distribute what they've uploaded and we trust them enough to push to stable". Maybe `is_approved_stable` instead? or just `is_approved`?

File: app/templates/user-admin.html (hunk 7 above)
Comment (hughsie) [reply app/templates/user-admin.html]: I think there are legitimate reasons why the admin or vendor admin remove attributes or change other things like the legacy username.

File: app/models.py (hunk 8 above)
Comment (superm1) [reply app/models.py]: Well I guess since they can do both testing and stable, I get what you're trying to convey.  Approved public makes enough sense.

File: app/templates/user-admin.html (hunk 7 above)
Comment (superm1) [reply app/templates/user-admin.html]: But so you can guard against the really bad stuff by hiding it then or turning a box grey if it's gonna be dangerous for example.  Like don't let them remove the admin attribute from themselves.

File: app/templates/firmware-affiliation.html [hunk 9]
```diff
@@ -13,6 +13,13 @@ <h2 class="mt-3">Vendor Affiliation</h2>
   owner the ablity to change the target and even delete the firmware.
 </p>
 
+{% if not g.user.check_acl('@admin') %}
+<div class="alert alert-warning mt-1" role="alert">
+  If you move this firmware to an different vendor it will not be possible
+  to edit the update details or move the firmware to testing or stable.
```
Comment (superm1) [inline app/templates/firmware-affiliation.html]: I think you should make it clear that the new owner can still do this.  Proposal:

"It will not be possible **for you** to edit the update details or move the firmware to testing or stable.  The new vendor will be able to perform those actions"

File: app/models.py [hunk 10]
```diff
@@ -727,6 +685,81 @@ def __init__(self):
         self.user_id = None         # user_id of the uploader
         self.mds = []
 
+    def check_acl(self, action, user=None):
```
Comment (superm1) [inline app/models.py]: I feel like you need a  helper function here for `is_owner(self, user)`.  Adding that you probably don't need most of the comments "is QA group and firmware owner" as it becomes more readable.

**Comment by superm1** [comment]
This is of course a huge PR so I haven't scrutinized super closely, but just a few small comments from me and it seems fine.

File: app/templates/user-admin.html (hunk 7 above)
Comment (hughsie) [reply app/templates/user-admin.html]: Right; checking this you can't actually "unadmin" yourself, so I've removed the all caps warning.

File: app/models.py (hunk 10 above)
Comment (hughsie) [reply app/models.py]: Good idea, fixed.
